File: backend/src/agents/ThinkingAgent.ts
import { BaseAgent } from './BaseAgent';
import { AgentResponse, AgentConfig } from './AgentInterface';
import { EventStream } from './EventStream';
import { LLMProvider } from './LLMProvider';
import { Action } from './Action';
import { Observation } from './Observation';
import langchainUtils from '../utils/langchain';

/**
 * Agent responsible for analyzing requests and creating execution plans
 */
export class ThinkingAgent extends BaseAgent {
  private thinkingChain;
  
  /**
   * Create a new ThinkingAgent
   * @param eventStream The event stream for the agent
   * @param llmProvider The LLM provider for the agent
   */
  constructor(eventStream: EventStream, llmProvider: LLMProvider) {
    super(eventStream, llmProvider);
    
    try {
      this.thinkingChain = langchainUtils.createThinkingAgentChain();
    } catch (error) {
      console.error('Error initializing ThinkingAgent chain:', error);
    }
  }
  
  /**
   * Initialize the agent
   * @param config Configuration for the agent
   */
  async initialize(config: AgentConfig): Promise<void> {
    await super.initialize(config);
    
    // Set default system message if not provided
    if (!this.systemMessage) {
      this.systemMessage = `You are the Thinking Agent, responsible for analyzing user requests and creating execution plans.
Your role is to break down complex tasks into smaller, manageable steps that other agents can execute.
You should consider the capabilities of other agents in the system when creating your plans.`;
    }
  }
  
  /**
   * Process an observation and generate an action
   * @param observation The observation to process
   * @returns The action to take
   */
  async process(observation: Observation): Promise<Action> {
    try {
      // Extract message from observation
      let message = '';
      let context = {};
      
      if (observation.data.type === 'user_message') {
        message = observation.data.message;
        context = observation.data.context || {};
      } else if (observation.data.type === 'text') {
        message = observation.data.content;
        context = observation.data.context || {};
      } else {
        // For other observation types, convert to string
        message = JSON.stringify(observation.data);
      }
      
      let plan: any;
      
      // Use LangChain if available, otherwise use mock plan
      if (this.thinkingChain) {
        try {
          const result = await this.thinkingChain.invoke({
            input: message,
          });
          
          // Parse the result as JSON
          plan = JSON.parse(result);
          
          // Add the original request to the plan
          plan.originalRequest = message;
        } catch (langchainError) {
          console.error('Error using LangChain:', langchainError);
          plan = this.createMockPlan(message);
        }
      } else {
        plan = this.createMockPlan(message);
      }
      
      // Create an action with the plan
      return new Action(this.id, {
        type: 'plan',
        content: plan
      });
    } catch (error) {
      console.error('Error in ThinkingAgent:', error);
      return Action.createErrorAction(this.id, 'Failed to analyze request and create plan');
    }
  }
  
  /**
   * Legacy method for backward compatibility
   * @param message The message to process
   * @param context Additional context
   * @returns A plan for executing the request
   * @deprecated Use process(observation) instead
   */
  async processMessage(message: string, context?: any): Promise<AgentResponse> {
    // Create an observation from the message
    const observation = Observation.createUserMessageObservation('user', message);
    
    // Add context to the observation if provided
    if (context) {
      observation.data.context = context;
    }
    
    // Process the observation
    const action = await this.process(observation);
    
    // Convert the action to a legacy response
    if (action.data.type === 'plan') {
      return {
        type: 'plan',
        content: action.data.content
      };
    } else {
      return this.actionToLegacyResponse(action);
    }
  }
  
  /**
   * Create a mock plan for demonstration purposes
   * @param message The user message
   * @returns A mock plan
   */
  private createMockPlan(message: string): object {
    return {
      originalRequest: message,
      analysis: "The user wants to create a simple web application.",
      steps: [
        {
          id: 1,
          description: "Create HTML structure",
          agent: "DeveloperAgent",
          params: { language: "html" }
        },
        {
          id: 2,
          description: "Add CSS styling",
          agent: "DeveloperAgent",
          params: { language: "css" }
        },
        {
          id: 3,
          description: "Implement JavaScript functionality",
          agent: "DeveloperAgent",
          params: { language: "javascript" }
        },
        {
          id: 4,
          description: "Review and improve code",
          agent: "EditorAgent",
          params: {}
        }
      ]
    };
  }
}

File: backend-fastapi/app/services/thinking_agent.py
"""
Thinking Agent for the NeuroNest AI system.
This agent analyzes requests and creates execution plans.
"""

import logging
import json
from typing import Dict, List, Any, Optional, Union

from app.services.base_agent import BaseAgent, EventStream, Action, Observation
from app.core.config import settings
from app.services.execution_service import execute_code_snippet
from langchain_openai import ChatOpenAI
from langchain.prompts import ChatPromptTemplate
from langchain.chains import LLMChain

logger = logging.getLogger(__name__)

class ThinkingAgent(BaseAgent):
    """
    Agent responsible for analyzing requests and creating execution plans
    """
    
    def __init__(self, event_stream: EventStream, llm_provider: Any):
        """
        Initialize a new ThinkingAgent
        
        Args:
            event_stream: The event stream for the agent
            llm_provider: The LLM provider for the agent
        """
        super().__init__(event_stream, llm_provider)
        self.name = "ThinkingAgent"
        self.description = "Analyzes requests and creates execution plans"
        self.capabilities = [
            "Analyze user requests",
            "Break down complex tasks",
            "Create structured execution plans",
            "Identify required agents and tools",
            "Prioritize tasks"
        ]
        self.thinking_chain = None
    
    async def initialize(self, config: Dict[str, Any]) -> None:
        """
        Initialize the agent
        
        Args:
            config: Configuration for the agent
        """
        await super().initialize(config)
        
        # Set default system message if not provided
        if not self.system_message:
            self.system_message = """You are the Thinking Agent, responsible for analyzing user requests and creating execution plans.
Your role is to break down complex tasks into smaller, manageable steps that other agents can execute.
You should consider the capabilities of other agents in the system when creating your plans.

Available agents:
- DeveloperAgent: Generates code in multiple languages
- EditorAgent: Reviews and improves code
- OrchestratorAgent: Coordinates other agents and tools
- ProjectPlanningAgent: Plans and structures software projects
- DebuggingAgent: Debugs code and fixes issues

Your output should be a JSON object with the following structure:
{
  "analysis": "Brief analysis of the user request",
  "steps": [
    {
      "id": 1,
      "description": "Step description",
      "agent": "AgentName",
      "params": { "param1": "value1", "param2": "value2" }
    },
    ...
  ]
}
"""
        
        # Initialize the thinking chain
        try:
            api_key = None
            if self.llm_provider and hasattr(self.llm_provider, "api_key"):
                api_key = self.llm_provider.api_key
            elif settings.OPENAI_API_KEY:
                api_key = settings.OPENAI_API_KEY
            
            if api_key:
                llm = ChatOpenAI(
                    model="gpt-4-turbo",
                    temperature=0.2,  # Lower temperature for more deterministic output
                    api_key=api_key
                )
                
                prompt = ChatPromptTemplate.from_messages([
                    ("system", self.system_message),
                    ("human", "{input}")
                ])
                
                self.thinking_chain = LLMChain(
                    llm=llm,
                    prompt=prompt,
                    output_key="output"
                )
        except Exception as e:
            logger.error(f"Error initializing thinking chain: {e}")
    
    async def process(self, observation: Observation) -> Action:
        """
        Process an observation and generate an action
        
        Args:
            observation: The observation to process
            
        Returns:
            The action to take
        """
        try:
            # Extract message from observation
            message = ""
            context = {}
            
            if observation.data["type"] == "user_message":
                message = observation.data["message"]
                context = observation.data.get("context", {})
            elif observation.data["type"] == "text":
                message = observation.data["content"]
                context = observation.data.get("context", {})
            else:
                # For other observation types, convert to string
                message = json.dumps(observation.data)
            
            # Generate plan
            plan = None
            
            # Use thinking chain if available
            if self.thinking_chain:
                try:
                    result = await self.thinking_chain.ainvoke({"input": message})
                    
                    # Try to parse the result as JSON
                    try:
                        plan_text = result["output"]
                        
                        # Extract JSON from the text if it's wrapped in markdown code blocks
                        if "```json" in plan_text and "```" in plan_text:
                            start = plan_text.find("```json") + 7
                            end = plan_text.find("```", start)
                            plan_text = plan_text[start:end].strip()
                        elif "```" in plan_text and "```" in plan_text:
                            start = plan_text.find("```") + 3
                            end = plan_text.find("```", start)
                            plan_text = plan_text[start:end].strip()
                        
                        plan = json.loads(plan_text)
                        
                        # Add the original request to the plan
                        plan["originalRequest"] = message
                    except json.JSONDecodeError:
                        logger.error(f"Error parsing JSON from thinking chain result: {result['output']}")
                        plan = self.create_mock_plan(message)
                except Exception as e:
                    logger.error(f"Error using thinking chain: {e}")
                    plan = self.create_mock_plan(message)
            else:
                plan = self.create_mock_plan(message)
            
            # Create an action with the plan
            return Action(self.id, {
                "type": "plan",
                "content": plan
            })
        except Exception as e:
            logger.error(f"Error in ThinkingAgent: {e}")
            return Action.create_error_action(self.id, f"Failed to analyze request and create plan: {str(e)}")
    
    def create_mock_plan(self, message: str) -> Dict[str, Any]:
        """
        Create a mock plan for demonstration purposes
        
        Args:
            message: The user message
            
        Returns:
            A mock plan
        """
        return {
            "originalRequest": message,
            "analysis": "The user wants to create a simple web application.",
            "steps": [
                {
                    "id": 1,
                    "description": "Create HTML structure",
                    "agent": "DeveloperAgent",
                    "params": { "language": "html" }
                },
                {
                    "id": 2,
                    "description": "Add CSS styling",
                    "agent": "DeveloperAgent",
                    "params": { "language": "css" }
                },
                {
                    "id": 3,
                    "description": "Implement JavaScript functionality",
                    "agent": "DeveloperAgent",
                    "params": { "language": "javascript" }
                },
                {
                    "id": 4,
                    "description": "Review and improve code",
                    "agent": "EditorAgent",
                    "params": {}
                }
            ]
        }